modules/pipeline/providers/leaderworker/leader.go

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderworker
+
+import (
+	"context"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+
+	"github.com/erda-project/erda/modules/pipeline/providers/leaderworker/worker"
+)
+
+func (p *provider) leaderFramework(ctx context.Context) {
+	p.Log.Infof("start leader framework")
+	defer p.Log.Infof("end leader framework")
+
+	// init before begin listen worker-task-change
+	p.initTaskWorkerAssignMap(ctx)
+
+	p.leaderUse.leaderHandlers = append(p.leaderUse.leaderHandlers, p.listenWorkerChange, p.continueCleanup, p.listenWorkerTaskIncoming, p.listenWorkerTaskDone)
+	p.workerUse.workerHandlersOnWorkerDelete = append(p.workerUse.workerHandlersOnWorkerDelete, p.workerIntervalCleanupOnDelete)
+
+	// leader
+	for _, h := range p.leaderUse.leaderHandlers {
+		h := h
+		go h(ctx)
+	}
+
+}
+
+func (p *provider) listenWorkerChange(ctx context.Context) {
+	// worker
+	notify := make(chan Event)
+	go func() {
+		for {
+			select {
+			case ev := <-notify:
+				switch ev.Type {
+				case mvccpb.PUT:
+					for _, h := range p.leaderUse.leaderHandlersOnWorkerAdd {
+						h := h
+						go h(ctx, ev)
```

Comment:
It depends on user.
At framework side, I have pass the cancellable context to function, it's enough.
On the other hand, worker won't change very frequently, because there only listen for official worker, and we have candidate worker mechanism to achieve that.

Reply:
maybe I should clarify worker/official-worker/candidate-worker.

Reply:
adjusted to `listenOfficialWorkerChange`.

Reply:
we do not care about candidate worker's change until it become official.